src/aws-cpp-sdk-core/include/smithy/tracing/TracingUtils.h

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#pragma once
+
+#include <aws/core/monitoring/HttpClientMetrics.h>
+#include <smithy/Smithy_EXPORTS.h>
+#include <smithy/tracing/Meter.h>
+#include <functional>
+#include <chrono>
+#include <utility>
+
+namespace smithy {
+    namespace components {
+        namespace tracing {
+            static const char SMITHY_METRICS_DNS_DURATION[] = "smithy.client.http.dns_duration";
+            static const char SMITHY_METRICS_CONNECT_DURATION[] = "smithy.client.http.connect_duration";
+            static const char SMITHY_METRICS_SSL_DURATION[] = "smithy.client.http.ssl_duration";
+            static const char SMITHY_METRICS_THROUGHPUT[] = "smithy.client.http.throughput";
+            static const char SMITHY_METRICS_UNKNOWN_METRIC[] = "smithy.client.http.unknown_metric";
+
+            class SMITHY_API TracingUtils {
+            public:
+                TracingUtils() = default;
+
+                template<typename T>
+                static T MakeCallWithTiming(std::function<T()> func,
+                    Aws::String metricName,
+                    std::shared_ptr<Meter> meter,
+                    Aws::Map<Aws::String, Aws::String> attributes,
+                    Aws::String description = "") {
+                    auto before = std::chrono::high_resolution_clock::now();
```

Comment:
nit:
there is a research doc by Mike Spertus about nasty compilers re-ordering instructions that can screw before/returnValue/after order of execution.

Reply:
also,
https://en.cppreference.com/w/cpp/chrono/high_resolution_clock

>Generally one should just use [std::chrono::steady_clock](https://en.cppreference.com/w/cpp/chrono/steady_clock) or [std::chrono::system_clock](https://en.cppreference.com/w/cpp/chrono/system_clock) directly instead of std::chrono::high_resolution_clock: use steady_clock for duration measurements, and system_clock for wall-clock time.

Reply:
good catch, changing to steady clock